lib/AST/CondBasedRefine.cpp

```diff
@@ -41,85 +29,67 @@ z3::expr CondBasedRefine::GetZ3Cond(clang::IfStmt *ifstmt) {
   return expr.simplify();
 }
 
-void CondBasedRefine::CreateIfThenElseStmts(IfStmtVec worklist) {
-  auto RemoveFromWorkList = [&worklist](clang::Stmt *stmt) {
-    auto it = std::find(worklist.begin(), worklist.end(), stmt);
-    if (it != worklist.end()) {
-      worklist.erase(it);
-    }
-  };
-
-  auto ThenTest = [this](z3::expr lhs, z3::expr rhs) {
-    return Prove(*z3_ctx, lhs == rhs);
-  };
-
-  auto ElseTest = [this](z3::expr lhs, z3::expr rhs) {
-    return Prove(*z3_ctx, lhs == !rhs);
-  };
-
-  auto CombineTest = [this](z3::expr lhs, z3::expr rhs) {
-    return Prove(*z3_ctx, z3::implies(rhs, lhs));
-  };
-
-  while (!worklist.empty()) {
-    auto lhs = *worklist.begin();
-    RemoveFromWorkList(lhs);
-    // Prepare conditions according to which we're going to
-    // cluster statements according to the whole `lhs`
-    // condition.
-    auto lcond = GetZ3Cond(lhs);
-    // Get branch candidates wrt `clause`
-    std::vector<clang::Stmt *> thens({lhs}), elses;
-    for (auto rhs : worklist) {
-      auto rcond = GetZ3Cond(rhs);
-      if (ThenTest(lcond, rcond) || CombineTest(lcond, rcond)) {
-        thens.push_back(rhs);
-      } else if (ElseTest(lcond, rcond) || CombineTest(!lcond, rcond)) {
-        elses.push_back(rhs);
-      }
-    }
+bool CondBasedRefine::VisitCompoundStmt(clang::CompoundStmt *compound) {
+  std::vector<clang::Stmt *> body{compound->body_begin(), compound->body_end()};
+  bool did_something{false};
+  for (size_t i{0}; i + 1 < body.size() && !Stopped(); ++i) {
+    auto if_a{clang::dyn_cast<clang::IfStmt>(body[i])};
+    auto if_b{clang::dyn_cast<clang::IfStmt>(body[i + 1])};
 
-    // Check if we have enough statements to work with
-    if (thens.size() + elses.size() < 2) {
+    if (!if_a || !if_b) {
       continue;
     }
 
-    // Erase then statements from the AST and `worklist`
-    for (auto stmt : thens) {
-      RemoveFromWorkList(stmt);
-      substitutions[stmt] = nullptr;
-    }
-    // Create our new if-then
-    auto sub = ast.CreateIf(lhs->getCond(), ast.CreateCompoundStmt(thens));
-    // Create an else branch if possible
-    if (!elses.empty()) {
-      // Erase else statements from the AST and `worklist`
-      for (auto stmt : elses) {
-        RemoveFromWorkList(stmt);
-        substitutions[stmt] = nullptr;
+    auto then_a{if_a->getThen()};
+    auto then_b{if_b->getThen()};
+
+    auto else_a{if_a->getElse()};
+    auto else_b{if_b->getElse()};
+
+    auto cond_a{GetZ3Cond(if_a)};
+    auto cond_b{GetZ3Cond(if_b)};
+
+    clang::IfStmt *new_if{};
+    if (Prove(*z3_ctx, cond_a == cond_b)) {
+      std::vector<clang::Stmt *> new_then_body{then_a, then_b};
```

Comment:
I think you can factor out this as well. I think everything besides that is fine.

Reply:
You mean `new_then_body`?